eip/EIPS/eip-4337.md

```diff
@@ -360,16 +360,76 @@ Stated in simpler terms, we expect at least `1 / MIN_INCLUSION_RATE_DENOMINATOR`
+
+In this proposal, we expect wallets to have a `validateUserOp` method that takes as input a `UserOperation`, and verify the signature and pay the fee. This method is required to be almost-pure: it is only allowed to access the storage of the wallet itself, cannot use environment opcodes (eg. `TIMESTAMP`), and can only edit the storage of the wallet, and can also send out ETH (needed to pay the entry point). The method is gas-limited by the `verificationGasLimit` of the `UserOperation`; nodes can choose to reject operations whose `verificationGasLimit` is too high. These restrictions allow miners and network nodes to simulate the verification step locally, and be confident that the result will match the result when the operation actually gets included into a block.
+
+The entry point-based approach allows for a clean separation between verification and execution, and keeps wallets' logic simple. The alternative would be to require wallets to follow a template where they first self-call to verify and then self-call to execute (so that the execution is sandboxed and cannot cause the fee payment to revert); template-based approaches were rejected due to being harder to implement, as existing code compilation and verification tooling is not designed around template verification.
+
+### Paymasters
+
+Paymasters facilitate transaction sponsorship, allowing third-party-designed mechanisms to pay for transactions. Many of these mechanisms _could_ be done by having the paymaster wrap a `UserOperation` with their own, but there are some important fundamental limitations to that approach:
+
+* No possibility for "passive" paymasters (eg. that accept fees in some ERC-20 token at an exchange rate pulled from an on-chain DEX)
+* Paymasters run the risk of getting griefed, as users could send ops that appear to pay the paymaster but then change their behavior after a block
 
-`eth_sendUserOperation`
+The paymaster scheme allows a contract to passively pay on users' behalf under arbitrary conditions. It even allows ERC-20 token paymasters to secure a guarantee that they would only need to pay if the user pays them: the paymaster contract can check that there is sufficient approved ERC-20 balance in the `validatePaymasterUserOp` method, and then extract it with `transferFrom` in the `postOp` call; if the op itself transfers out or de-approves too much of the ERC-20s, the inner `postOp` will fail and revert the execution and the outer `postOp` can extract payment (note that because of storage access restrictions the ERC-20 would need to be a wrapper defined within the paymaster itself).
```

Comment:
With the new rule in AA-67 we can access the user's balance in 3rd party ERC20 contracts, as long as the `balanceOf` function of that token doesn't access other storage slots not associated with the user's address.  However, this doesn't make TokenPaymaster completely safe because it can't check the token's exchange rate on Uniswap or anything like it.  It could use different strategies to protect itself:
* If it's a frequently used paymaster, it can save a copy of the price during `postOp`, and in each validation it requires the user to have at least 2x the required amount based on the saved price, to account for price fluctuations.
* It could be a DepositPaymaster, requiring the user to lock a security deposit once, and ask the user to provide the Uniswap price of the token as part of the paymasterData.  If the paymaster discovers (during `postOp` that the user lied and doesn't have sufficient balance+allowance, it forfeits the security deposit to cover the damage.